scanpy/tools/_ingest.py

```diff
@@ -0,0 +1,239 @@
+            self._rep = adata.obsm['X_pca'][:, :N_PCS]
+            self._n_pcs = self._rep.shape[1]
+
+        if 'metric_kwds' in adata.uns['neighbors']['params']:
+            dist_args = tuple(adata.uns['neighbors']['params']['metric_kwds'].values())
+        else:
+            dist_args = ()
+        dist_func = named_distances[adata.uns['neighbors']['params']['metric']]
+        self._random_init, self._tree_init = make_initialisations(dist_func, dist_args)
+        self._search = make_initialized_nnd_search(dist_func, dist_args)
+
+        search_graph = adata.uns['neighbors']['distances'].copy()
+        search_graph.data = (search_graph.data > 0).astype(np.int8)
+        self._search_graph = search_graph.maximum(search_graph.transpose())
+
+        if 'rp_forest' in adata.uns['neighbors']:
+            self._rp_forest = _rp_forest_generate(adata.uns['neighbors']['rp_forest'])
+        else:
+            self._rp_forest = None
+
+    def _init_pca(self, adata):
+        self._pca_centered = adata.uns['pca']['params']['zero_center']
+        self._pca_use_hvg = adata.uns['pca']['params']['use_highly_variable']
+
+        if self._pca_use_hvg and 'highly_variable' not in adata.var.keys():
+            raise ValueError('Did not find adata.var[\'highly_variable\'].')
+
+        if self._pca_use_hvg:
+            self._pca_basis = adata.varm['PCs'][adata.var['highly_variable']]
+        else:
+            self._pca_basis = adata.varm['PCs']
+
+    def __init__(self, adata):
+        #assume rep is X if all initializations fail to identify it
+        self._rep = adata.X
+        self._use_rep = 'X'
+
+        self._n_pcs = None
+
+        self._adata_ref = adata
+        self._adata_new = None
+
+        if 'pca' in adata.uns:
+            self._init_pca(adata)
+
+        if 'neighbors' in adata.uns:
+            self._init_neighbors(adata)
+
+        if 'X_umap' in adata.obsm:
+            self._init_umap(adata)
+
+        self._obsm = None
+        self._obs = None
+        self._labels = None
+
+        self._indices = None
+        self._distances = None
+
+    def _pca(self, n_pcs=None):
+        X = self._adata_new.X
+        X = X.toarray() if issparse(X) else X.copy()
+        if self._pca_use_hvg:
+            X = X[:, self._adata_ref.var['highly_variable']]
+        if self._pca_centered:
+            X -= X.mean(axis=0)
+        X_pca = np.dot(X, self._pca_basis[:, :n_pcs])
+        return X_pca
+
+    def _same_rep(self):
+        adata = self._adata_new
+        if self._n_pcs is not None:
+            return self._pca(self._n_pcs)
+        if self._use_rep == 'X':
+            return adata.X
+        if self._use_rep in adata.obsm.keys():
+            return adata.obsm[self._use_rep]
+        return adata.X
+
+    def transform(self, adata_new):
+        self._obs = pd.DataFrame(index=adata_new.obs.index)
+        #not sure if it should be AxisArrays
+        self._obsm = AxisArrays(adata_new, 0)
```

Comment:
I don't think it does, really that class is mostly to support indexing of AnnData objects, but this class can't be subset. I suppose it could be used for the shape checks, but that's easy enough to do separately.

Reply:
I found it to be convenient here, maybe i need to get rid of it because it is an internal AnnData thing.